### ISSUE TITLE
[Request] Sponsor Blocking

### ISSUE THREAD

**Issue by illeatmyhat** [TeamNewPipe/NewPipe#2554]
There's a project using a crowd-sourced tagging database to skip over parts of the video dedicated to sponsorship and advertisements.
https://github.com/ajayyy/SponsorBlock

I think that it could be an interesting idea to integrate.

**Comment by frstrikerman** [comment]
I just take a look at the API and it should be simple to get timecodes I will look on making a script for mpv if everything work as expected I will try to do the same here

**Comment by Z3ROGRAV1TYx** [comment]
+1 for this, love this on my computer

**Comment by polymorphicshade** [comment]
This would be great in NewPipe.
I've added API calls to the source code and successfully got videos to skip sponsors automatically.
Below are some previews of the changes I've made on my local machine.
If y'all think this would fit nicely in NewPipe, let me know and I will send a pull request.
![01_](https://user-images.githubusercontent.com/32504507/69501093-bc109580-0ebe-11ea-8ce0-ca5444c19fee.jpg)
![02_](https://user-images.githubusercontent.com/32504507/69501094-bc109580-0ebe-11ea-862c-27885bb8cedf.jpg)
![03_](https://user-images.githubusercontent.com/32504507/69501095-bc109580-0ebe-11ea-8452-e0490cbf4af4.jpg)

**Comment by Z3ROGRAV1TYx** [comment]
@polymorphicshade nice, I think there should be a feature to hide button for those who want

**Comment by polymorphicshade** [comment]
@Z3ROGRAV1TYx yeah currently if that setting is disabled, the icon will be hidden from the video player

**Comment by illeatmyhat** [comment]
LGTM. Does it highlight the labeled seekbar segments? Is there an undo? 
To throw more options at the wall, a different way of doing it would be to have two buttons for "segment start" and "segment end" at the current playback time, so that they could be paired together. One button wouldn't work because it would be easy to mess up.

**Comment by polymorphicshade** [comment]
@illeatmyhat basically I plan to mimic the browser plugin as much as I can, the screenshots above are just something I threw together in a day or so.
I agree having something like a toggle button to mark the start/end of sponsor times would be better than a single-button popup dialog.

**Comment by lionirdeadman** [comment]
Personally, I don't think this is something that NewPipe should do considering we already can't support creators through Youtube's ads and most people probably won't just give money directly to all the creators they like since that's not really economically viable.

But, this also gives a precedent to altering the video the creators make intentionally and creating blacklists of segments of videos. I think modifying or blacklisting content (or parts of content) is not a rabbit-hole that NewPipe should participate in.

**Comment by polymorphicshade** [comment]
@lionirdeadman I disagree. I believe more in giving viewers control on whether or not they view ads/sponsors than succumbing to intrusive ads. If the viewer wants to support the creator, they can simply not use this feature.
Of course this is up to the NewPipe devs (or whoever approves merging changes), but I believe due to its open source nature, it should be up to the general public to decide if they want to use/enable a feature or not.

**Comment by aloisdg** [comment]
@lionirdeadman  I want to see the fall of Youtube. I hope to see creators moving to a less capitalist greey platform. Peertube+Liberapay perhabs? Right now one of the main attractive point of NewPipe is to skip ads. See this feature as an extension of this main point.

**Comment by lionirdeadman** [comment]
@aloisdg Well, that's what you think is one of the main attractive point. Currently, Newpipe doesn't show ads because it's impossible, not out of choice. For me, Newpipe is simply what I can use to watch videos without the proprietary Youtube application.

And sure, it would be great if Peertube + a payment service were to succeed Youtube but it's not by removing bread from people's mouth that it should be done, that's just unethical at best.

**Comment by Alexander--** [comment]
The idea of blocking ads at video segment level sounds all right to me.

But the proposed approach (using SponsorBlock API) is bad for user privacy, because it allows SponsorBlock owners to learn exactly which video you watch and when.

I'd prefer if NewPipe downloaded entire database with timestamps (it is not big, — AFAIK, just few megabytes right now) and used it to block ads without consulting SponsorBlock servers.

**Comment by ajayyy** [comment]
@Alexander-- SponsorBlock is just made by me. But yea, it for sure can't be default. A local database would probably be significantly more difficult to implement and wouldn't be as real-time.

**Comment by aloisdg** [comment]
@lionirdeadman when I talk to non tech-savvy people about NewPipe, the native adblocking is a better selling point that the freedom of the software.

@Alexander-- Downloading the db to access it offline could be optional.

**Comment by Z3ROGRAV1TYx** [comment]
> 
> 
> @Alexander-- SponsorBlock is just made by me. But yea, it for sure can't be default. A local database would probably be significantly more difficult to implement and wouldn't be as real-time.

As an optional feature implemented though, it would be great.

**Comment by lionirdeadman** [comment]
> @lionirdeadman when I talk to non tech-savvy people about NewPipe, the native adblocking is a better selling point that the freedom of the software.

Well that's sad and well, somewhat untrue, we don't do any adblocking to my knowledge, we just get the video file and that doesn't have any ads to begin with.

**Comment by katktv** [comment]
I personally don't want to waste *any* time on advertisements, whether native or sponsored, and given the complete lack of regulations in this regard we will absolutely get yet another NordVPN fiasco that will damage many more people bc they were featured in videos of someone they watch. It already got worse in Russia where many youtubers started advertising stuff like gambling, real or CS GO skin one, shady G2A-esque game stores, and so on. 

By skipping ads I am protecting myself from the wild west that is the internet. It ain't like creators won't get paid from me not watching the ads they did not put in the video.

**Comment by aloisdg** [comment]
@lionirdeadman Alright my term is quite misleading. NewPipe doesnt use an adblocking system but is ad-free.

**Comment by Z3ROGRAV1TYx** [comment]
> @lionirdeadman Alright my term is quite misleading. NewPipe doesnt use an adblocking system but is ad-free.

Personally, I found your statement very clear. NewPipe isn't using an adblocking system, it just ONLY retrieves the videos, and the ads do not come alongside it. I don't think this needs to be argued for anymore. Ads in general only support channels very little anyways. Waste your time, etc. 99% of people use an adblocker now adays anyways. Personally I will go out of my way to block every single ad on a page if I can. They take of space, bloat the page, can introduce scummy tracking techniques, etc. They ruin the internet.

This is not the point of this issue here. I think Sponsorblock implementation would be great though, it would be optional so if people had an issue with disabling sponsors then they'd keep it disabled. If they want sponsor messages, they enable it. Sponsor messages imo are the biggest waste of my time, and I skip over them anyways. On the browser I use sponsorblock anyways. I would never buy something personally simply because of someone promoting something, if anything makes me buy it less. I do my own research. So they are useless to me. I hate them wasting my time, Sponsorblock just makes the process of skipping them 100x easier and providing a more seamless watching experience. 

So I'm hoping it gets implemented in the near future and I'm sure many other people do as well.

**Comment by thesuperzapper** [comment]
@polymorphicshade any update on when you’ll be ready to do a PR?

Or do you want any help?

**Comment by polymorphicshade** [comment]
> 
> 
> @polymorphicshade any update on when you’ll be ready to do a PR?
> 
> Or do you want any help?

I have most of want I want done; I'm just looking into something else for a bit until I create a PR. Probably will make one some time this weekend.

**Comment by KonoromiHimaries** [comment]
https://github.com/TeamNewPipe/NewPipe/issues/1819

**Comment by illeatmyhat** [comment]
Are we supposed to break out the folding chairs and fight now?

**Comment by grravity** [comment]
Skimmed the PR, so it seems like we are close to having this merged and officially supported soon? Or are we quite a ways away?

**Comment by polymorphicshade** [comment]
> 
> 
> Skimmed the PR, so it seems like we are close to having this merged and officially supported soon? Or are we quite a ways away?

I wouldn't say it's close to being merged, I'm waiting on a much bigger PR to be merged before I continue working on fixing the progress bar overlay. However aside from simply skipping ads automatically, I'm not sure what other SponsorBlock-specific features would be worth putting time in to, considering NewPipe's intent on being a light front-end for YouTube.

**Comment by grravity** [comment]
SponsorBlock does have a few more things it skips now that may be worth considering, not sure how much more weight it'd put on Newpipe, maybe it should be good if they're disabled?

**Comment by Piruzzolo** [comment]
Since NewPipe reads the page html, would it be possibile to load the ads but not showing them, somehow fooling YT ads system?

**Comment by Atemu** [comment]
That's an entirely different different topic, please create a new issue.

**Comment by Piruzzolo** [comment]
So is this issue's idea would be better then https://github.com/TeamNewPipe/NewPipe/issues/4437 ? I can't see how, blocking ads is already depriving creators of earnings, but you guys seem to have no problem with this.

**Comment by polymorphicshade** [comment]
@Piruzzolo think of it this way: if you watch TV and the channel doesn't show ads to you, but tricks the ad companies into thinking you actually did see the ads, then this is essentially theft.

Instead, if the channel lets you decide to watch or skip an ad, and the ad company knows about this, then it's ok.

**Comment by ajayyy** [comment]
@Piruzzolo That's a great way to get the creator of the video's you are watching banned from adsense.

**Comment by TheAssassin** [comment]
Team NewPipe decided not to support SponsorBlock in NewPipe. Team NewPipe doesn't want to be a gatekeeper, deciding which forms of advertising we want to allow in the app, and which we don't want to allow. We've taken the time to write down our reasoning in detail. Please see [our position paper](https://newpipe.schabi.org/blog/pinned/newpipe-and-online-advertising/) for more information.

**Comment by grravity** [comment]
That's very sad to hear, especially seeing Vanced has introduced this feature. I fail to see the issue seeing it would be toggleable so users who wish to see sponsors can. They could also choose to use their own lists. Sponsorblock already offers multiple toggles for different forms of advertising so if users want some but not others they can, or just choose to have all. I think this will hold us back, and just continue to be a pain to the users who already manually skip the sponsors anyways. With all due respect, I personally don't see this as a good decision.

**Comment by NotAFile** [comment]
Thank you for the position paper. I agree with the general sentiment that newpipe should not block ethical options for advertising by default.

However I don't quite see the "gatekeeper" function you described in the feature. Current versions of SponsorBlock allow users to tag sections of the video for many other things including intro sequences. It also lets users choose between a number of options such as showing a skip button or just showing it in the seek bar.

Since this is provided by other applications, I feel that the net effect is to reduce the viability of using free software options over proprietary solutions that offer more features.

**Comment by GarbageHamburger** [comment]
While I agree with the paper, SponsorBlock does not in any way affect the creators who get sponsorships. As far as I can tell, creators get paid a fixed sum as soon as the video with the sponsorship is published, and whether users skip the sponsor segments do not in any way affect the creator. As @NotAFile stated, SponsorBlock also allows users to selectively skip (or skip by clicking a button) so the gatekeeper argument doesn't seem to be valid.
Please reconsider support for this feature as it is really useful and doesn't seem to be detrimental to anybody.

**Comment by Ra-Na** [comment]
I think SponsorBlock is an excellent idea and enjoy polymorphicsshade fork! Please keep it up.

**Comment by JasonHK** [comment]
SponsorBlock not just skip ads only, but also intro, outro and subscription reminder etc.

There are some **authorized** anime distributors (Ani-One Asia, Muse Asia) that release full-length anime episodes to YouTube. Other viewers and I submitted the intro and outro of those animes the distributors released so others could skip those not-so-important parts.
